src/Agent/NewRelic/Agent/Core/Transactions/Transaction.cs

```diff
@@ -59,9 +59,15 @@ public ISegment CurrentSegment
                     return Segment.NoOpSegment;
                 }
 
-                if (Segments[currentSegmentIndex.Value] != null)
+                int idx = currentSegmentIndex.Value;
+                if (idx >= Segments.Count)
                 {
-                    return Segments[currentSegmentIndex.Value];
+                    Log.Warn($"Transaction {Guid} is out of sync with the current segment [looking for {idx} out of {Segments.Count}]");
```

Comment:
Do we want this to be a warn or a debug level message?  Warns will show up in info level logs.

Reply:
I think that this should be rare enough that it should be ok at the warn level. It will also make it easier for us to be aware that it is happening. I think something similar is done for transactions that are captured by finalier/destructor.

Reply:
It was intentional, unless we have guidelines for what constitutes a Warning? Although I fixed the exception, if this happens it means these Segments are being associated with the wrong Transaction, which is pretty bad.